### PR TITLE
Make it compatible with Adapt 4 and Metal 1

### PR DESCRIPTION
**Pull request by joelreymont**
Prevent RL from downgrading Metal 1 on the Mac, enabling Lux to use the Mac GPU.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -1,6 +1,9 @@
 name = "ReinforcementLearning"
 uuid = "158674fc-8238-5cab-b5ba-03dfc80d1318"
-authors = ["Johanni Brea <jbrea@users.noreply.github.com>", "Jun Tian <tianjun.cpp@gmail.com>"]
+authors = [
+    "Johanni Brea <jbrea@users.noreply.github.com>",
```
Comment (jeremiahpslewis) [inline Project.toml:4]: Please revert the changes to this file, at the moment the umbrella package isn't being shipped with the latest versions of the dependencies.

File: src/ReinforcementLearningDatasets/Project.toml [hunk 2]
```diff
@@ -22,24 +22,25 @@ ReinforcementLearningEnvironments = "25e41dd2-4622-11e9-1641-f1adca772921"
...
 TFRecord = "841416d8-1a6a-485a-b0fc-1328d0f53d5e"
 UnicodePlots = "b8865327-cd53-5732-bb35-84acbb429228"
+cuDNN = "02a925ec-e4fe-4b08-9a7e-0d78e3d38ccd"
 
 [compat]
 CodecZlib = "0.7"
 DataDeps = "0.7"
 Flux = "0.13, 0.14"
-HDF5 = "0.15"
+HDF5 = "0.15, 0.17"
```
Comment (jeremiahpslewis) [inline src/ReinforcementLearningDatasets/Project.toml:31]: what about `0.16`?

File: src/ReinforcementLearningZoo/Project.toml [hunk 3]
```diff
@@ -30,7 +30,7 @@ LogExpFunctions = "0.3"
 NNlib = "0.8, 0.9"
 Optim = "1"
 ReinforcementLearningBase = "0.12"
-ReinforcementLearningCore = "0.15.1"
+ReinforcementLearningCore = "0.15"
```
Comment (jeremiahpslewis) [inline src/ReinforcementLearningZoo/Project.toml:33]: please drop this, there was a change to rlcore that means we want at least the `0.15.1` version

File: src/ReinforcementLearningDatasets/Project.toml [hunk 4]
```diff
@@ -22,24 +22,25 @@ ReinforcementLearningEnvironments = "25e41dd2-4622-11e9-1641-f1adca772921"
...
+Pickle = "0.3"
 Pipe = "1.3"
 ProgressMeter = "1.7"
 PyCall = "1.92"
-ReinforcementLearningBase = "0.9"
-ReinforcementLearningEnvironments = "0.6.3"
-Setfield = "0.7"
-TFRecord = "0.3"
-UnicodePlots = "1.3"
+ReinforcementLearningBase = "0.12"
```
Comment (jeremiahpslewis) [inline src/ReinforcementLearningDatasets/Project.toml:39]: currently, rldatasets is unmaintained and not included in the ci/cd, please drop the changes to this file...

File: Project.toml (hunk 1 above)
Comment (joelreymont) [reply Project.toml:4]: This causes issues when I try to compile the umbrella package to fail to pre-compile. 

If I try to dev the sub-packages I get 

```
(ReinforcementLearning) pkg> dev src/ReinforcementLearningZoo src/ReinforcementLearningEnvironments src/ReinforcementLearningCore src/ReinforcementLearningBase
   Resolving package versions...
ERROR: empty intersection between ReinforcementLearningZoo@0.10.1 and project compatibility 0.6
```

What do you suggest?

**Comment by joelreymont** [comment]
Here's why I updated the umbrella package and undertook updating RL in the first place.

Lux was at 0.5.15 and Metal at 1.0 but the outdated umbrella package versions downgrade both Lux and Metal.

More importantly, Lux does not use the GPU (Metal) on the Mac when downgraded but does use the GPU with Metal 1.0. 

Is there a way for me to use RL _and_ my Mac GPU with Lux while keeping old RL sub-packages in the umbrella RL package?

```
(@v1.10) pkg> st
Status `~/.julia/environments/v1.10/Project.toml`
⌃ [b2108857] Lux v0.5.15
⌃ [dde4c033] Metal v0.5.0
  [158674fc] ReinforcementLearning v0.11.0-dev `~/Work/Julia/ReinforcementLearning.jl`
⌃ [02a925ec] cuDNN v1.1.1
Info Packages marked with ⌃ have new versions available and may be upgradable.

(@v1.10) pkg> add Metal@1
   Resolving package versions...
ERROR: Unsatisfiable requirements detected for package Adapt [79e6a3ab]:
 Adapt [79e6a3ab] log:
 ├─possible versions are: 0.3.0-4.0.1 or uninstalled
 ├─restricted by compatibility requirements with Metal [dde4c033] to versions: 4.0.0-4.0.1
 │ └─Metal [dde4c033] log:
 │   ├─possible versions are: 0.0.1-1.0.0 or uninstalled
 │   └─restricted to versions 1 by an explicit requirement, leaving only versions: 1.0.0
 └─restricted by compatibility requirements with ReinforcementLearningCore [de1b191a] to versions: 3.0.0-3.7.2 — no versions left
   └─ReinforcementLearningCore [de1b191a] log:
     ├─possible versions are: 0.1.0-0.13.0 or uninstalled
     └─restricted to versions 0.9 by ReinforcementLearning [158674fc], leaving only versions: 0.9.0-0.9.4
       └─ReinforcementLearning [158674fc] log:
         ├─possible versions are: 0.11.0 or uninstalled
         └─ReinforcementLearning [158674fc] is fixed to version 0.11.0-dev
```

**Comment by joelreymont** [comment]
Everything works fine if the umbrella package is updated. Meaning that RL
compiles and does not downgrade metal. I will re-test and I dropped the
changes you requested.

I plan to actively use RL and can put time into it. I’m new to Julia but
I’m learning quickly.

**Comment by jeremiahpslewis** [comment]
As a first step, can you get the CI to pass?

Side note: for the time being, it's best to install the sub packages separately and NOT include the umbrella package RL.jl in your environment. This is also how the GitHub Actions CI is set up.

**Comment by joelreymont** [comment]
Let me take a look…

**Comment by joelreymont** [comment]
Some tests are failing and `OpenSpiel` dependency is failing to compile.

I'm on it.

**Comment by jeremiahpslewis** [comment]
ok, a note, I need to tag and release new versions of RLTrajectories and CircularArrays before they are picked up here.

**Comment by joelreymont** [comment]
> ok, a note, I need to tag and release new versions of RLTrajectories and CircularArrays before they are picked up here.

How does this affect me?

**Comment by jeremiahpslewis** [comment]
I think this is necessary...
- https://github.com/JuliaPackaging/Yggdrasil/pull/8219
- https://github.com/JuliaRegistries/General/pull/101984#issuecomment-1971186549

**Comment by joelreymont** [comment]
Issue #1010 prevents CI from passing.

**Comment by jeremiahpslewis** [comment]
this will fix it, need to wait for GitHub https://github.com/JuliaReinforcementLearning/OpenSpiel.jl/pull/19

**Comment by jeremiahpslewis** [comment]
once this merges, rlenv is hopefully good to go: https://github.com/JuliaRegistries/General/pull/102066

**Review by jeremiahpslewis — APPROVED** [review]

**Comment by joelreymont** [comment]
What's the next step?

**Comment by jeremiahpslewis** [comment]
I’ll release the new versions of RLCore / RLEnv. Can you track down the bug in the offlineagents?

**Comment by joelreymont** [comment]
I'm on it.
